File: src/scoreboard-client/lib/dtos/api/eval_records.ts
export interface EvalRecord {
  sample_index: number;
  repeat_index: number;
  pass_index: number;
  is_passed: boolean;
  answer: string;
  ref_answer: string;
  fail_reason: string;
  context_preview?: string;
}

export interface EvalRecordsResponse {
  task_id: number;
  records: EvalRecord[];
  offset: number;
  limit: number;
  next_offset: number;
  has_more: boolean;
}

File: src/scoreboard-server/scoreboard_server/services/api/eval_records.py
from __future__ import annotations

from scoreboard_server.db.repository import ScoreboardStore
from scoreboard_server.dtos.api.eval_records import EvalRecordsResponse


async def eval_records_response(
    store: ScoreboardStore,
    *,
    task_id: int,
    only_wrong: bool,
    limit: int,
    offset: int,
) -> EvalRecordsResponse:
    records = await store.list_eval_records_for_space(
        task_id=str(task_id),
        only_wrong=only_wrong,
        limit=limit + 1,
        offset=offset,
        include_context=False,
        include_preview=True,
    )
    page = records[:limit]
    return {
        "task_id": task_id,
        "records": page,
        "offset": offset,
        "limit": limit,
        "next_offset": offset + len(page),
        "has_more": len(records) > limit,
    }
